### PR TITLE
[HOPSWORKS-3127] Manage and monitor resource usage for Docker containers

### PR DESCRIPTION
**Pull request by gibchikafa**

### CODE REVIEW

File: recipes/docker_registry.rb [hunk 1]
```diff
@@ -201,3 +201,5 @@
     action :systemd_reload
   end
 end
+
+include_recipe "hops::docker_cgroup"
```
Comment (jimdowling) [inline recipes/docker_registry.rb:205]: if exists_local("hopsworks", "default") 
  include_recipe "hops::docker_cgroup"
end

File: recipes/docker.rb [hunk 2]
```diff
@@ -221,3 +221,5 @@
 kagent_config service_name do
   action :systemd_reload
 end
+
+include_recipe "hops::cadvisor"
```
Comment (jimdowling) [inline recipes/docker.rb:225]: if exists_local("hopsworks", "default")
  include_recipe "..."
end

File: recipes/docker_cgroup.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+docker_cgroup_cpu_cfs_quota_us = -1
+docker_cgroup_memory_hard_limit = 9223372036854771712
+docker_cgroup_memory_soft_limit = 9223372036854771712
+if node['hops']['docker']['cgroup']['enabled'].eql?("true")
+  docker_cgroup_memory_hard_limit = "#{node['hops']['docker']['cgroup']['memory']['hard-limit']}"
```
Comment (jimdowling) [inline recipes/docker_cgroup.rb:5]: For a "dynamic" way to set memory based on available memory in the host, do this:

create in attributes.default

node['hops']['docker']['cgroup']['memory']['hard-limit-default'] = 6
node['hops']['docker']['cgroup']['memory']['hard-limit'] = node['hops']['docker']['cgroup']['memory']['hard-limit-default']

node['hops']['docker']['cgroup']['memory']['soft-limit-default'] = 4
node['hops']['docker']['cgroup']['memory']['soft-limit'] = node['hops']['docker']['cgroup']['memory']['soft-limit-default']



Then in docker_cgroup.rb

available_memory_in_host_gbs = node.memory.total[/\d*/].to_i
default_hard_memory = node['hops']['docker']['cgroup']['memory']['hard-limit']
default_soft_memory = node['hops']['docker']['cgroup']['memory']['soft-limit']

available_memory_in_host_gbs = available_memory_in_host_gbs - default_hard_memory
excess_memory = available_memory_in_host_gbs - 30
excess_memory = excess_memory / 4

  # if the user overrides the hard-limit, then we should use their new hard-limit, and not add excess_memory
if node['hops']['docker']['cgroup']['memory']['hard-limit-default'] != node['hops']['docker']['cgroup']['memory']['hard-limit']
  excess_memory = 0
end

if node['hops']['docker']['cgroup']['memory']['soft-limit-default'] != node['hops']['docker']['cgroup']['memory']['soft-limit']
  excess_memory = 0
end

node.override['hops']['docker']['cgroup']['memory']['hard-limit'] = default_hard_memory + excess_memory
node.override['hops']['docker']['cgroup']['memory']['soft-limit'] = default_soft_memory + excess_memory

**Review by jimdowling — CHANGES_REQUESTED** [review]

File: attributes/default.rb [hunk 4]
```diff
@@ -535,3 +535,16 @@
...
+default['hops']['cadvisor']['download-url']        = "#{node['download_url']}/docker/cadvisor"
+default['hops']['cadvisor']['port']                = "4194"
+
+
+default['hops']['docker']['cgroup']['enabled']                       = "true"
+default['hops']['docker']['cgroup']['memory']['hard-limit-default']  = "6GB"
+default['hops']['docker']['cgroup']['memory']['hard-limit']          = node['hops']['docker']['cgroup']['memory']['hard-limit-default']
+default['hops']['docker']['cgroup']['memory']['soft-limit-default']  = "2GB"
+default['hops']['docker']['cgroup']['memory']['soft-limit']          = node['hops']['docker']['cgroup']['memory']['soft-limit-default']
+default['hops']['docker']['cgroup']['cpu']['quota']                  = 70
```
Comment (robzor92) [inline attributes/default.rb:549]:
```suggestion
default['hops']['docker']['cgroup']['cpu']['quota']                  = 80
```

File: metadata.rb [hunk 5]
```diff
@@ -955,6 +955,25 @@
           :description =>  "Directory to mount the filesystem",
           :type => 'string'
 
+<<<<<<< HEAD
 attribute "hops/yarn/nodemanager-graceful-decommission-timeout-secs",
           :description => "This is the maximal time to wait for running containers and applications to complete before transition a DECOMMISSIONING node into DECOMMISSIONED. The default value is 3600 seconds. Negative value (like -1) is handled as infinite timeout.",
           :type => 'string'
+=======
```
Comment (robzor92) [inline metadata.rb:962]:
```suggestion
```

File: metadata.rb [hunk 6]
```diff
@@ -955,6 +955,25 @@
           :description =>  "Directory to mount the filesystem",
           :type => 'string'
 
+<<<<<<< HEAD
```
Comment (robzor92) [inline metadata.rb:958]:
```suggestion
```

File: metadata.rb [hunk 7]
```diff
@@ -955,6 +955,25 @@
...
+
+attribute "hops/docker/cgroup/memory/soft-limit",
+          :description => "Soft limit in bytes to allow flexible sharing of memory.  When the system detects memory contention or low memory, containers will restrict their consumption to the soft limits.",
+          :type => 'string'
+
+attribute "hops/docker/cgroup/cpu/quota",
+          :description => "The percentage of total cpu consumption for docker containers.",
+          :type => 'string'
+
+>>>>>>> Manage and monitor resource usage for Docker containers
```
Comment (robzor92) [inline metadata.rb:979]:
```suggestion
```

File: attributes/default.rb [hunk 8]
```diff
@@ -535,3 +535,16 @@
...
+default['hops']['docker']['cgroup']['cpu']["period"]                 = 100000
```
Comment (robzor92) [inline attributes/default.rb:550]:
```suggestion
default['hops']['docker']['cgroup']['cpu']['period']                 = 100000
```

File: attributes/default.rb [hunk 9]
```diff
@@ -535,3 +535,16 @@
...
 default['hops']['nn']['audit_log']                 = "false"
 default['hops']['rm']['audit_log']                 = "false"
+
+default['hops']['cadvisor']['dir']                 = "#{node['hops']['dir']}/cadvisor"
+default['hops']['cadvisor']['download-url']        = "#{node['download_url']}/docker/cadvisor"
+default['hops']['cadvisor']['port']                = "4194"
+
+
+default['hops']['docker']['cgroup']['enabled']                       = "true"
+default['hops']['docker']['cgroup']['memory']['hard-limit-default']  = "6GB"
```
Comment (robzor92) [inline attributes/default.rb:545]: Why do you have a _default['hops']['docker']['cgroup']['memory']['hard-limit-default']_ and a _default['hops']['docker']['cgroup']['memory']['hard-limit']_, shouldn't it be enough with one of them?

Comment (robzor92) [reply attributes/default.rb:545]: Like only keep default['hops']['docker']['cgroup']['memory']['hard-limit'] and just set it to 6GB.

File: recipes/cadvisor.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  variables({
+              'cadvisor_bin' => cadvisor_bin
+            })
+end
+
+kagent_config "cadvisor" do
+  action :systemd_reload
+end
+
+=begin
```
Comment (robzor92) [inline recipes/cadvisor.rb:53]: Why is this commented out?

File: recipes/docker_cgroup.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  excess_memory = 0
+end
+excess_memory = (excess_memory/4).round()
+
+node.override['hops']['docker']['cgroup']['memory']['hard-limit'] = ((docker_hard_limit_memory_bytes + excess_memory)/1073741824).to_s + "GB"
+node.override['hops']['docker']['cgroup']['memory']['soft-limit'] = ((docker_soft_limit_memory_bytes + excess_memory) /1073741824).to_s + "GB"
+
+if node['hops']['docker']['cgroup']['enabled'].eql?("true")
+  cpu_quota_value = node['hops']['docker']['cgroup']['cpu']['quota']
+  cpu_quota_period = node['hops']['docker']['cgroup']['cpu']["period"]
```
Comment (robzor92) [inline recipes/docker_cgroup.rb:19]:
```suggestion
  cpu_quota_period = node['hops']['docker']['cgroup']['cpu']['period']
```

File: attributes/default.rb (hunk 8 above)
Comment (robzor92) [reply attributes/default.rb:550]: Also can you add this attribute to metadata.rb?

**Review by robzor92 — CHANGES_REQUESTED** [review]

File: recipes/docker_cgroup.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
+available_memory_in_host_bytes = (node['memory']['total'].gsub("kB","").to_i/10241024) * 1073741824
+docker_default_hard_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['hard-limit-default'].gsub("GB", "").to_i) * 1073741824
+docker_dedault_soft_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['soft-limit-default'].gsub("GB", "").to_i) * 1073741824
```
Comment (robzor92) [inline recipes/docker_cgroup.rb:3]:
```suggestion
docker_default_soft_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['soft-limit-default'].gsub("GB", "").to_i) * 1073741824
```

File: recipes/docker_cgroup.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+docker_hard_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['hard-limit'].gsub("GB", "").to_i) * 1073741824
+docker_soft_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['soft-limit'].gsub("GB", "").to_i) * 1073741824
+# dyanamic memory allocation
+available_memory_in_host_bytes = available_memory_in_host_bytes - docker_hard_limit_memory_bytes
+excess_memory = available_memory_in_host_bytes - (30 * 1073741824)
+if excess_memory < 0 || (docker_default_hard_limit_memory_bytes != docker_hard_limit_memory_bytes || docker_dedault_soft_limit_memory_bytes != docker_soft_limit_memory_bytes)
```
Comment (robzor92) [inline recipes/docker_cgroup.rb:9]:
```suggestion
if excess_memory < 0 || (docker_default_hard_limit_memory_bytes != docker_hard_limit_memory_bytes || docker_default_soft_limit_memory_bytes != docker_soft_limit_memory_bytes)
```

File: recipes/docker_cgroup.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
+available_memory_in_host_bytes = (node['memory']['total'].gsub("kB","").to_i/10241024) * 1073741824
+docker_default_hard_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['hard-limit-default'].gsub("GB", "").to_i) * 1073741824
+docker_dedault_soft_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['soft-limit-default'].gsub("GB", "").to_i) * 1073741824
+docker_hard_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['hard-limit'].gsub("GB", "").to_i) * 1073741824
+docker_soft_limit_memory_bytes = (node['hops']['docker']['cgroup']['memory']['soft-limit'].gsub("GB", "").to_i) * 1073741824
+# dyanamic memory allocation
```
Comment (robzor92) [inline recipes/docker_cgroup.rb:6]:
```suggestion
# dynamic memory allocation
```

File: recipes/docker_cgroup.rb [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+available_memory_in_host_bytes = available_memory_in_host_bytes - docker_hard_limit_memory_bytes
+excess_memory = available_memory_in_host_bytes - (30 * 1073741824)
```
Comment (robzor92) [inline recipes/docker_cgroup.rb:8]: Add a comment that the assumption is that Hopsworks services require 30GB.

File: templates/default/cadvisor.service.erb [hunk 16]
```diff
@@ -0,0 +1,27 @@
+[Unit]
+Description = cadvisor
+StartLimitIntervalSec=30
+StartLimitBurst=20
+
+[Service]
+ExecStart = <%= @cadvisor_bin %> \
+--storage_driver_host=localhost:8086 \
+--storage_driver_password=root \
```
Comment (robzor92) [inline templates/default/cadvisor.service.erb:9]: Is it a security risk by exposing this password here? Are users able to connect to it and what information can they get?

**Review by robzor92 — CHANGES_REQUESTED** [review]

**Review by robzor92 — APPROVED** [review]

**Review by jimdowling — APPROVED** [review]

**Pull request merged**